thetis/callback.py

```diff
@@ -417,8 +375,61 @@ def minmax():
             tracer_min = self.solver_obj.comm.allreduce(tracer_min, op=MPI.MIN)
             tracer_max = self.solver_obj.comm.allreduce(tracer_max, op=MPI.MAX)
             return tracer_min, tracer_max
-        super(TracerOvershootCallBack, self).__init__(minmax,
-                                                      solver_obj,
-                                                      outputdir=outputdir,
-                                                      export_to_hdf5=export_to_hdf5,
-                                                      append_to_log=append_to_log)
+        super(TracerOvershootCallBack, self).__init__(minmax, solver_obj, **kwargs)
+
+
+class DetectorsCallback(DiagnosticCallback):
+    """Callback that writes the specified fields interpolated in the specified detector locations"""
+    name = 'detectors'
+
+    def __init__(self, solver_obj,
+                 detector_locations,
+                 field_names,
+                 detector_names=None,
+                 **kwargs):
+        """
+        :arg solver_obj: Thetis solver object
+        :list detector_locations: Locations in which fields are to be interpolated.
+        :list field_names: Fields to be interpolated.
+        :list detector_names: Names for each of the detectors. If not provided automatic names
+           of the form 'detectorNNN' are created where NNN is the (0-padded) detector number.
+        :arg **kwargs: any additional keyword arguments, see DiagnosticCallback
+        """
+        # printing all detector output to log is probably not a useful default:
+        kwargs.setdefault('append_to_log', False)
+        field_dims = [solver_obj.fields[field_name].function_space().value_size
+                          for field_name in field_names]
+        attrs = {
+                 # use null-padded ascii strings, dtype='U' not supported in hdf5, see http://docs.h5py.org/en/latest/strings.html
+                 'field_names': np.array(field_names, dtype='S'),
+                 'field_dims': field_dims,
+                }
+        super().__init__(solver_obj, array_dim=sum(field_dims), attrs=attrs, **kwargs)
+
+        ndetectors = len(detector_locations)
+        if detector_names is None:
+            fill = len(str(ndetectors))
+            detector_names = ['detector' + str(i).zfill(fill) for i in range(ndetectors)]
+        else:
+            assert ndetectors == len(detector_names), "Different number of detector locations and names"
+        self._variable_names = detector_names
```

Comment:
Hmm semantically variable names should be related to the field names? I.e. a variable is a physical quantity with a unit. (In the long run these hf5 output files should be CF compliant in some sense...)

Reply:
I was just trying to reuse the existing DiagnosticHDF5. I'd rather not have `f=h5py.File(); data=f['fieldname']` and `data` be a array of ntimesteps x ndetectors, i.e. not use detector names but just have them in one big array in the specified order. We typically have a large number of detectors and it gets hard to keep track of which ones, and in which order, have been used in a specific run. In particular because this tend to keep changing for subsequent runs in a specific domain as we change the mesh slightly (moving boundaries) and add more data sources. We could have `f['fieldname']['detectorname'][:]` if you prefer (or the other way around)?

Reply:
Yes, I see your point. I guess this is fine as it is. If we did `f['detectorname']['fieldname'][:]` we could attach CF metadata to the individual arrays, but at the moment `DiagnosticHDF5` class doesn't support such nesting and I'm not sure if that would be easy to implement.